jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/internal/HttpStreamOverHTTP2.java

```diff
@@ -237,8 +248,7 @@ public Runnable onTrailer(HeadersFrame frame)
         HttpFields trailers = frame.getMetaData().getHttpFields().asImmutable();
         try (AutoLock ignored = lock.lock())
         {
-            _demand = false;
-            _chunk = new Trailers(trailers);
+            _trailer = new Trailers(trailers);
```

Comment:
We know that end stream has not already been read?  I guess we could not read EOF unless trailers had been delivered?

Reply:
`DATA` frames are accumulated in the `Stream` and then read, but the trailer is notified immediately, hence the race condition this PR tries to fix.

When this method is called we know that all `DATA` frames have been accumulated, but they may have not been read yet, so we only remember the trailer here, and figure out the reads in `read()`.

Reply:
My concern is the other way around. i.e. can all the data frames be read before (or while) this method is called?  What is the guarantee that we have not already returned an EOF chunk to the application before we get the trailers.

My guess of how it is implemented makes me think that it will be OK, but would be good to confirm exactly why and say so in a comment.

Reply:
It might be worth mentioning in a comment that when a trailers `HEADERS` frame gets parsed, the stream enqueues a `EOF` data before calling `HttpStreamOverHTTP2.onTrailer()`.